src/software/amazon/ionschema/internal/constraint/Regex.kt

```diff
@@ -0,0 +1,61 @@
+package software.amazon.ionschema.internal.constraint
+
+import software.amazon.ion.IonString
+import software.amazon.ion.IonText
+import software.amazon.ion.IonValue
+import software.amazon.ionschema.InvalidSchemaException
+import software.amazon.ionschema.Violations
+import software.amazon.ionschema.Violation
+import software.amazon.ionschema.internal.CommonViolations
+import javax.script.ScriptEngineManager
+
+/**
+ * Implements the regex constraint.
+ *
+ * @see https://amzn.github.io/ion-schema/docs/spec.html#regex
+ */
+internal class Regex(
+        ion: IonValue
+) : ConstraintBase(ion) {
+
+    companion object {
+        private val scriptEngine = ScriptEngineManager().getEngineByName("javascript")
+    }
+
+    private val regex = if (ion !is IonString || ion.isNullValue) {
+            throw InvalidSchemaException("Invalid regex constraint: $ion")
+        } else {
+            ion.stringValue()
+        }
+
+    private val flags: String
+
+    init {
+        val sb = StringBuffer()
+        ion.typeAnnotations.forEach {
+            when (it) {
+                "i", "m" -> sb.append(it)
+                else -> throw InvalidSchemaException(
+                        "Unrecognized flags for regex ($ion)")
+            }
+        }
+        flags = sb.toString()
+    }
+
+    override fun validate(value: IonValue, issues: Violations) {
```

Comment:
The `validate()` method structure of:
* Is it the right data type?
* Is it null when it shouldn't be?
* Operator-specific validation

also seems to be very prevalent pattern.

Reply:
Refactored by introducing `ConstraintBase.validateAs()` which performs the type and null checks, then passes the value (cast to the expected type) to a custom validation block.  I'll admit to appreciating the simplicity of seeing the full validation logic in each constraint impl, but agree that the duplication would eventually lead to inconsistencies.

Note that ConstraintBaseIntRange (and the restriction that class type parameters can't be reified) is the reason for the validateAs() variant with `expectedClass` as the first parameter.

This change simplifies Contains, Element, Regex, TimestampOffset, and TimestampPrecision impls (plus Fields and OrderedElements constraints, which are outside the scope of this PR).